RecoTracker/MkFit/plugins/MkFitOutputConverter.cc

```diff
@@ -0,0 +1,580 @@
+        // inactive hits
+        const auto* detLayer = detLayers.at(hitOnTrack.layer);
+        if (detLayer == nullptr) {
+          throw cms::Exception("LogicError") << "DetLayer for layer index " << hitOnTrack.layer << " is null!";
+        }
+        // Actually it is necessary to leave dealing with invalid hits to the TrackProducer?
+        //recHits.push_back(new InvalidTrackingRecHitNoDet(detLayer->surface(), TrackingRecHit::missing)); // let's put them all as missing for now
+        lastHitInvalid = true;
+      } else {
+        recHits.push_back(indexLayers.getHitPtr(hitOnTrack.layer, hitOnTrack.index)->clone());
+        LogTrace("MkFitOutputConverter") << "  pos " << recHits.back().globalPosition().x() << " "
+                                         << recHits.back().globalPosition().y() << " "
+                                         << recHits.back().globalPosition().z() << " mag2 "
+                                         << recHits.back().globalPosition().mag2() << " detid "
+                                         << recHits.back().geographicalId().rawId() << " cluster "
+                                         << indexLayers.getClusterIndex(hitOnTrack.layer, hitOnTrack.index);
+        lastHitInvalid = false;
+      }
+    }
+
+    const auto lastHitId = recHits.back().geographicalId();
+
+    // MkFit hits are *not* in the order of propagation, sort by 3D radius for now (as we don't have loopers)
+    // TODO: Improve the sorting (extract keys? maybe even bubble sort would work well as the hits are almost in the correct order)
+    recHits.sort([](const auto& a, const auto& b) {
+      const auto aid = a.geographicalId();
+      const auto bid = b.geographicalId();
+
+      const auto asub = aid.subdetId();
+      const auto bsub = bid.subdetId();
+      if (asub != bsub) {
+        // Subdetector order (BPix, FPix, TIB, TID, TOB, TEC) corresponds also the navigation
+        return asub < bsub;
+      }
+
+      const auto& apos = a.globalPosition();
+      const auto& bpos = b.globalPosition();
+
+      if (isBarrel(asub)) {
+        return apos.perp2() < bpos.perp2();
+      }
+      return std::abs(apos.z()) < std::abs(bpos.z());
+    });
+
+    const bool lastHitChanged = (recHits.back().geographicalId() != lastHitId);  // TODO: make use of the bools
+
+    // seed
+    const auto seedIndex = cand.label();
+    LogTrace("MkFitOutputConverter") << " from seed " << seedIndex << " seed hits";
+    const auto& mkseed = mkfitSeeds.at(cand.label());
+    for (int i = 0; i < mkseed.nTotalHits(); ++i) {
+      const auto& hitOnTrack = mkseed.getHitOnTrack(i);
+      LogTrace("MkFitOutputConverter") << "  hit on layer " << hitOnTrack.layer << " index " << hitOnTrack.index;
+      // sanity check for now
+      const auto& candHitOnTrack = cand.getHitOnTrack(i);
+      if (hitOnTrack.layer != candHitOnTrack.layer) {
+        //throw cms::Exception("LogicError")
```

Comment:
delete commented-out code?

Reply:
Let me check if I could turn the LogErrors back to exceptions (this situation better not happen, but as you can see, it has happened in the past).

Reply:
With 100k 10-mu and 10k ttbar nopu/pu50/pu70 the error condition didn't occur, so I turned the LogErrors back to exceptions.